Ogs documentation (#156)
jbathmann/pyMANGA: web/content/en/docs/Example_OGS_BETTINA/The_Groundwater_Domain.md

    # Pressure at a given point. Here, pressure is 0 at the surface

    def ini_pressure_function(point):
        return -(1000 * 9.81 * (point[2] - transectElevation(x=point[0])))

    # Concentration at a given point. Here, c_ini is constant 0.035 kg/kg.

    def c(point):
        return .035

The following function later adds the defined pressure and concentration field to the domain:

    def addCandPtoMesh(mesh):
        points = mesh.GetPoints()
        # Preparing lists for initial conditions
        p_ini = vtk.vtkDoubleArray()
        p_ini.SetName("p_ini")
        c_ini = vtk.vtkDoubleArray()
        c_ini.SetName("c_ini")
        # Calculating initial conditions
        for point in range(points.GetNumberOfPoints()):
            point = points.GetPoint(point)
            c_ini.InsertNextTuple1(c(point))
            p_ini.InsertNextTuple1(-(1000 * 9.81 *
                                   (point[2] -
                                      transectElevation(x=point[0],
                                                      m=hydraulic_gradient))))
        mesh.GetPointData().AddArray(p_ini)
        mesh.GetPointData().AddArray(c_ini)

> Also, the following block can be used to create a cuboid lattice.

Reviewer: ```suggestion
The following block can be used to create a cuboid lattice.
```

> The grid is divided into three layers with the layer thicknesses *l_z_top*, *l_z_mid* and *l_z_bottom*.

Author: ```suggestion
The grid is divided into three layers with the layer thicknesses *l_z_top*, *l_z_mid* and *l_z_bottom*.
If one of the lengths is set to zero, the layer is not created.
```

A vertical resolution can be specified for each of these layers (*num_top*, *num_bottom*, *num_mid*).
The *z* parameter specifies a possible displacement of the top edge of the terrain, the basic course of which is defined in *TransectElevation*, in meters.

> *two_layers* and *three_layers* defines how many different layers there are.

Author: ```suggestion
```

        lines.append(geom.add_line(points[2], points[1]))
        lines.append(geom.add_line(points[1], points[0]))

        line_loop = geom.add_curve_loop(lines)
        surface = geom.add_plane_surface(line_loop)

        baum_1 = [20, 5, transect_elevation(20, m=m) + z]
        p_baum_1 = geom.add_point(baum_1, lcar / 2)
        geom.in_surface(p_baum_1, surface)

        baum_2 = [22.5, 5, transect_elevation(22.5, m=m) + z]
        p_baum_2 = geom.add_point(baum_2, lcar / 2)
        geom.in_surface(p_baum_2, surface)
        return surface

    ## Generates a mesh for given parameters
    #  @param z: shift in z-direction
    #  @param l_z_top: depth of the top layer (if only one layer created, value
    #  defines the depth)
    #  @param l_z_bottom: depth of the bottom layer (if two layers created, value
    #  defines the depth of the bottom layer)
    #  @param num_top: resolution of top layer
    #  @param num_bottom: resolution of bottom layer
    #  @param l_y: y-extension of grid
    #  @param l_z_mid: depth of mid layer, if constructed
    #  @param num_mid: resolution of mid layer
    #  @param l_x: x-extension of model

>     #  @param lcar: characteristic lengthscale in x-direction

Author: ```suggestion
    #  @param lcar: characteristic lengthscale in x- and y-direction in meters 
```

    #  @param transect_elevation: Function, which returns a z-value for given x-val

>     def meshGen(z,
>                 l_z_top,
>                 l_z_bottom=0,
>                 num_top=1,
>                 num_bottom=0,
>                 l_y=10,
>                 l_z_mid=0,
>                 num_mid=0,
>                 l_x=230,
>                 lcar=5,
>                 transect_elevation=transectElevation,
>                 m=1e-3):

Author: ```suggestion
    def meshGen(z,
                l_z_top,
                l_z_mid=0,
                l_z_bottom=0,
                num_top=1,
                num_mid=0,
                num_bottom=0,
                l_y=10,
                l_x=230,
                lcar=5,
                transect_elevation=transectElevation,
                m=1e-3):
```

    propertyvector.SetName("permeability")
    bulk_cells = vtk.vtkCellArray()
    for cell_point_ids in bulky.cells[2].data:
        cell = vtk.vtkTetra()
        for i in range(4):
            cell.GetPointIds().SetId(i, cell_point_ids[i])

        propertyvector.InsertNextTuple1(3.0e-11)

        bulk_cells.InsertNextCell(cell)
    bulk.SetCells(10, bulk_cells)
    bulk.GetCellData().AddArray(propertyvector)

Since some points were created twice during mesh creation, the vtkStaticCleanUnstructuredGrid filter must be applied to merge them.
Then the property "bulk_node_ids", which is necessary for OGS, is inserted.

    # Clean the data with non-zero tolerance
    clean1 = vtk.vtkStaticCleanUnstructuredGrid()
    clean1.SetInputData(bulk)
    clean1.ToleranceIsAbsoluteOff()
    clean1.SetTolerance(0.0001)
    clean1.RemoveUnusedPointsOn()

    clean1.Update()
    bulk = clean1.GetOutput()
    propertyvector = vtk.vtkDataArray.CreateDataArray(
        vtk.VTK_UNSIGNED_LONG_LONG)
    propertyvector.SetName("bulk_node_ids")
    for i in range(bulk.GetNumberOfPoints()):
        propertyvector.InsertNextTuple1(i)
    bulk.GetPointData().AddArray(propertyvector)

> Then the initial pressure field and the initial concentration field. have been added, the bulk mesh can be saved.

Reviewer: ```suggestion
When the initial pressure field and the initial concentration field have been added, the bulk mesh can be saved.
```

    for i in range(ncells):
        cells.GetNextCell(pts)
        if pts.GetNumberOfIds() == 4:
            p0 = np.array(points.GetPoint(pts.GetId(0)))
            p1 = np.array(points.GetPoint(pts.GetId(1)))
            p2 = np.array(points.GetPoint(pts.GetId(2)))
            p3 = np.array(points.GetPoint(pts.GetId(3)))
            v0 = (p1 - p0)
            v1 = (p2 - p0)
            v2 = (p3 - p0)
            V = np.absolute(np.dot(v0, np.cross(v1, v2))) / 6.
            cell_volumes[i] = V

To copy the properties we assigned to the main domain, vtk's *ResampleWithDataset* filter is used.

    resample_filter = vtk.vtkResampleWithDataSet()
    resample_filter.SetSourceData(bulk)
    resample_filter.SetInputData(source)
    resample_filter.Update()
    source = resample_filter.GetOutput()

    for value in cell_volumes:
        propertyvector.InsertNextTuple1(value)
    source.GetCellData().AddArray(propertyvector)

    # Output of source-mesh
    source_writer = vtk.vtkXMLUnstructuredGridWriter()
    source_writer.SetFileName("my_first_source.vtu")
    source_writer.SetInputData(source)
    source_writer.Write()

At the end only the boundary meshes are missing.

> The ExtractSurface utility from the opengeosys project can be used to extract the boundary mesh.

Reviewer: ```suggestion
Finally, we create the boundary meshes using the ExtractSurface utility from the OGS project.
```
